File: src/main/java/com/reddit/recommender/ingestion/ScraperParameters.java
package com.reddit.recommender.ingestion;

import org.json.JSONArray;
import org.json.JSONObject;
import java.util.List;
import java.util.ArrayList;

public class ScraperParameters {
    private JSONObject parameters;

    public ScraperParameters() {
        this.parameters = new JSONObject();
        // Set defaults
        parameters.put("maxPosts", 10);
        parameters.put("scrapeComments", false);
        parameters.put("includeNsfw", false);
        parameters.put("sort", "relevance");
        parameters.put("timeframe", "month");
    }

    // Builder pattern methods

    public ScraperParameters addQuery(String query) {
        if (!parameters.has("queries")) {
            parameters.put("queries", new JSONArray());
        }
        parameters.getJSONArray("queries").put(query);
        return this;
    }

    public ScraperParameters addQueries(List<String> queries) {
        if (!parameters.has("queries")) {
            parameters.put("queries", new JSONArray());
        }
        for (String query : queries) {
            parameters.getJSONArray("queries").put(query);
        }
        return this;
    }

    public ScraperParameters addUrl(String url) {
        if (!parameters.has("urls")) {
            parameters.put("urls", new JSONArray());
        }
        parameters.getJSONArray("urls").put(url);
        return this;
    }

    public ScraperParameters addUrls(List<String> urls) {
        if (!parameters.has("urls")) {
            parameters.put("urls", new JSONArray());
        }
        for (String url : urls) {
            parameters.getJSONArray("urls").put(url);
        }
        return this;
    }

    public ScraperParameters setMaxPosts(int maxPosts) {
        parameters.put("maxPosts", maxPosts);
        return this;
    }

    public ScraperParameters setMaxComments(int maxComments) {
        parameters.put("maxComments", maxComments);
        return this;
    }

    public ScraperParameters setScrapeComments(boolean scrapeComments) {
        parameters.put("scrapeComments", scrapeComments);
        return this;
    }

    public ScraperParameters setIncludeNsfw(boolean includeNsfw) {
        parameters.put("includeNsfw", includeNsfw);
        return this;
    }

    public ScraperParameters setSort(String sort) {
        // Valid values: "relevance", "hot", "top", "new", "comments"
        parameters.put("sort", sort);
        return this;
    }

    public ScraperParameters setTimeframe(String timeframe) {
        // Valid values: "hour", "day", "week", "month", "year", "all"
        parameters.put("timeframe", timeframe);
        return this;
    }

    // Advanced parameters
    public ScraperParameters setProxyConfiguration(JSONObject proxyConfig) {
        parameters.put("proxyConfiguration", proxyConfig);
        return this;
    }

    public ScraperParameters setCustomData(JSONObject customData) {
        parameters.put("customData", customData);
        return this;
    }

    // Clear methods
    public ScraperParameters clearQueries() {
        if (parameters.has("queries")) {
            parameters.remove("queries");
        }
        return this;
    }

    public ScraperParameters clearUrls() {
        if (parameters.has("urls")) {
            parameters.remove("urls");
        }
        return this;
    }

    // Build method
    public JSONObject build() {
        return parameters;
    }

    // Static factory methods for common scenarios

    public static JSONObject createSearchParameters(String query, int maxPosts) {
        return new ScraperParameters()
                .addQuery(query)
                .setMaxPosts(maxPosts)
                .build();
    }

    public static JSONObject createSubredditParameters(String subreddit, int maxPosts) {
        String url = "https://www.reddit.com/r/" + subreddit + "/";
        return new ScraperParameters()
                .addUrl(url)
                .setMaxPosts(maxPosts)
                .build();
    }

    public static JSONObject createAdvancedParameters(List<String> queries, List<String> urls,
                                                      int maxPosts, boolean scrapeComments,
                                                      String sort, String timeframe) {
        ScraperParameters builder = new ScraperParameters()
                .setMaxPosts(maxPosts)
                .setScrapeComments(scrapeComments)
                .setSort(sort)
                .setTimeframe(timeframe);

        if (queries != null) {
            builder.addQueries(queries);
        }

        if (urls != null) {
            builder.addUrls(urls);
        }

        return builder.build();
    }
}

File: src/main/scala/com/reddit/recommender/api/ScraperParameters.scala
package com.reddit.recommender.api

import org.json.{JSONArray, JSONObject}
import java.util.{List => JList}
import scala.collection.JavaConverters._
class ScraperParameters {
  private val parameters = new JSONObject()

//default values
// those values are for mot depleting my credit
  parameters.put("maxPosts", 10)
  parameters.put("scrapeComments", false)
  parameters.put("includeNsfw", false)
  parameters.put("sort", "relevance")
  parameters.put("timeframe", "week")
  parameters.put("maxComments", 5)   




  //search for words or sentences
  def addQuery(query: String): ScraperParameters = {
    if (!parameters.has("queries")) {
      parameters.put("queries", new JSONArray())
    }
    parameters.getJSONArray("queries").put(query)
    this
  }

  def addQueries(queries: JList[String]): ScraperParameters = {
    if (!parameters.has("queries")) {
      parameters.put("queries", new JSONArray())
    }
    queries.asScala.foreach(parameters.getJSONArray("queries").put)
    this
  }
 // direct Reddit URL to scrape like post or subreddit
  def addUrl(url: String): ScraperParameters = {
    if (!parameters.has("urls")) {
      parameters.put("urls", new JSONArray())
    }
    parameters.getJSONArray("urls").put(url)
    this
  }

  def addUrls(urls: JList[String]): ScraperParameters = {
    if (!parameters.has("urls")) {
      parameters.put("urls", new JSONArray())
    }
    urls.asScala.foreach(parameters.getJSONArray("urls").put)
    this
  }
// max number of posts for the query on the api
  def setMaxPosts(maxPosts: Int): ScraperParameters = {
    parameters.put("maxPosts", maxPosts)
    this
  }

  def setMaxComments(maxComments: Int): ScraperParameters = {
    parameters.put("maxComments", maxComments)
    this
  }
//do I want comments or not?
  def setScrapeComments(scrapeComments: Boolean): ScraperParameters = {
    parameters.put("scrapeComments", scrapeComments)
    this
  }

  // this is only here as a concept, im not going to use this,
  // for posts that are regarded as not safe for work,
  // things that contain mature content
  // like violent videos are marked as not safe for work
   def setIncludeNsfw(includeNsfw: Boolean): ScraperParameters = {
     parameters.put("includeNsfw", includeNsfw)
     this
   }
  /** Accepted values:
   *   "relevance" , best match to query (default)
   *   "new"       , newest first
   *   "top"       , highest score
   *   "hot"       , trending
   *   "controversial" */
  def setSort(sort: String): ScraperParameters = {
    parameters.put("sort", sort)
    this
  }
  /** Sets time range for posts
   * Accepted values:
   *   "hour"
   *   "day"
   *   "week" (default)
   *   "month"
   *   "year"
   *   "all"
   */
  def setTimeframe(timeframe: String): ScraperParameters = {
    parameters.put("timeframe", timeframe)
    this
      }




  def build(): JSONObject = parameters
}

object ScraperParameters {
  def createSearchParameters(query: String, maxPosts: Int): JSONObject = {
    new ScraperParameters()
      .addQuery(query)
      .setMaxPosts(maxPosts)
      .build()


  }}